### ISSUE TITLE
pid 1 caught ABRT, dumped core, freezing execution; while stopping now aliased service

### ISSUE THREAD

**Issue by BtbN** [systemd/systemd#9315]
**systemd version the issue has been seen with**
> 238

**Used distribution**
 > Gentoo

**Expected behaviour you didn't see**

Well, I expected pid 1 not to abort and freeze.

**Unexpected behaviour you saw**
 > Broadcast message from systemd-journald@btbn.de (Sat 2018-06-16 11:21:41 CEST):
 > systemd[1]: Caught &lt;ABRT>, dumped core as pid 16151.
 > Broadcast message from systemd-journald@btbn.de (Sat 2018-06-16 11:21:41 CEST):
 >systemd[1]: Freezing execution.

After that the system was in need of a forceful reboot.

I have a core dump file of the incident:
https://btbn.de/files/core.systemd.0.e9d15573f5214f228e955136bc8b5212.16151.1529140901000000.lz4
Not sure if it contains any sensitive information, if so, please let me know and I'll pull it down again.

**Steps to reproduce the problem**

What I suspect to be the cause of this is me trying to migrate from mysqld.service to mariadb.service.
mariadb.service provides an alias for mysqld.service.
So I disabled mysqld.service and enabled mariadb.service. That created symlinks in /etc/systemd, called mysqld.service pointing to mariadb.service, as it's expected for aliased services.
Then I wanted to stop the old mysqld.service, and start mariadb.service.
But after issuing the stop command, other systemctl commands started hanging, and shortly after I got the above broadcasts.

**Comment by keszybz** [comment]
On Sat, Jun 16, 2018 at 03:16:53AM -0700, BtbN wrote:
> **systemd version the issue has been seen with**
> > 238
> 
> **Used distribution**
>  > Gentoo

Hi,

unfortunately most (all?) systemd developers don't use gentoo, and
even if, gentoo packages are compiled locally with differing settings,
so the core is only useful on that local system. You'll have to
generate a backtrace yourself. Please un-lz4 the core file, and point
gdb to it with 'gdb /lib/systemd/systemd core.systemd.0.e9d15573f5214f228e955136bc8b5212.16151.1529140901000000', and do a 'bt full' and attach that.

Zbyszek

**Comment by BtbN** [comment]
I don't have any debugging symbols, so the backtrace is not overly useful:
```
#0  0x00007f0acc0142c7 in ?? ()
No symbol table info available.
#1  0x00005567e797a487 in ?? ()
No symbol table info available.
#2  0x0000000000000000 in ?? ()
No symbol table info available.
```

**Comment by keszybz** [comment]
Not much we can do in this case. If you can reproduce the error, try compiling with debug symbols and provide a backtrace.

**Comment by trofi** [comment]
How to get debugging symbols in Gentoo: https://wiki.gentoo.org/wiki/Debugging

**Comment by BtbN** [comment]
This must be a bad joke. I'm not going to try crash a production server again.

The description of this issue seems clear enough to me to at least have a rough idea where things might go wrong.
Gentoo is probably using one of the most vanilla systemd builds compared to most distributions...

There must be an issue with handling still running now aliased services somewhere that causes an abort in pid1.

**Comment by poettering** [comment]
@BtbN we don't doubt that there is a bug somewhere. But it's not really possible for us to track this down with a backtrace of some form. We'd love to investigate this, but we simply can't without debug symbols.

Note that systemd tries to make it easy to keep a system debuggable, through "coredumpctl" and friends. systemd will automatically generate backtraces of all coredumps as they happen and make them easily accessible. But this all breaks down if debug symbols are not available, as this means coredumps become just a big pile of bits, without metadata to make sense of them.

Distributions such as Fedora nowadays always ship with minimal debug symbols to ensure that a system is always debuggable to some point. But if a system is stripped of all of this, this really nothing we can do. 

Sorry.

**Comment by filbranden** [comment]
For the record:

```
systemd[1]: Caught <ABRT>, dumped core as pid 16151.
systemd[1]: Freezing execution.
```

This is normal, if the system manager on PID 1 gets a SIGABRT. You can easily reproduce this by running this as root:

```
# kill -ABRT 1
```

The part that's hard for us to reproduce is figuring out what might have sent a SIGABRT to pid 1... Which also is probably not really in the core dump file you have (I imagine it will start with the receiving of the SIGABRT and then dumping/freezing from there.)

**Comment by BtbN** [comment]
So even if I were to be able to reproduce this, which on a second attempt replaying the commands from history didn't happen again, it would be hard/impossible to track back?

**Comment by kirbyfan64** [comment]
Basically, you would first have to have a systemd with debug symbols (which I'm pretty sure you can do on Gentoo). That would give *far* more information and hopefully at least a rough idea of where the problem is. Right now, it would be pretty much impossible to figure out.

**Comment by trofi** [comment]
> This must be a bad joke. I'm not going to try crash a production server again.

You don't have to. As long as you build the same source with the same flags (with `-ggdb` or similar) enabled you should be able to decode your existing `core` file.

**Comment by BtbN** [comment]
> You don't have to. As long as you build the same source with the same flags (with -ggdb or similar) enabled you should be able to decode your existing core file.

That's what I tried initially. It made gdb complain that core file and binary do not match up, and made it produce that broken backtrace.

This doesn't seem to be easily reproducible. At least reverting to mysqld.service, and doing the switch again the same way I did originally did not produce any problems this time.
For the future I have systemd and glibc built with debug symbols, so should it ever happen again, there will hopefully be more info available.
